File: pathFinder.cc
#include "pathFinder.h"
#include "car.h"
#include "road.h"
#include <algorithm>
#include <iterator>
#include <set>
#include <memory>
#include <stack>
#include "constants.h"
#include "openElement.h"

PathFinder::PathFinder(const Ground &g, int(*funcH)(int, int, int, int)):
	g{g}, funcH{funcH} {}

PathFinder::~PathFinder() {}

std::queue<int> PathFinder::findPath(int x, int y, int desX, int desY, Car *car) {
	std::queue<int> container;
	// pair.first is f score; pair.second is Tile
	// invariant: sorted in increasing order
	// format: <fScore, gScore, current Tile *, previous openElement ptr, direction>
	auto firstNode = std::make_unique<OpenElement>(funcH(x, y, desX, desY), 0, car->getRoad(), nullptr, DEFAULT);
	OpenContainer open;
	open.emplace_back(std::move(firstNode));
	std::set<std::unique_ptr<OpenElement>> closed;
	
	// local variables
	int fScore, gScore;
	Tile *t;
	Tile *tile;
	int dir;
	int curX, curY;
	std::vector<Tile *> neighbours;
	OpenElement *prev;

	// child variables
	int fScoreChild, gScoreChild;
	int childX, childY;
	int childRelX, childRelY;
	
	//// DEV TEST ////
	int X, Y;
	std::string s;

	while(!open.empty()) {
		// selects the front of open set (which has lowest f-score based on invariant)
		auto node = std::move(open.front());
		OpenElement *nodePtr = node.get();
		open.pop_front();
		std::tie(fScore, gScore, t, prev, dir) = node->pack();
		closed.insert(std::move(node));
		
		std::tie(X, Y) = t->getCoord();
		std::cout << "New openElement selected: " << '(' << X << ',' << Y << ')' << std::endl;
		std::getline(std::cin, s);


		std::cout << "Current chain:" << std::endl;
		OpenElement *testNode = nodePtr;
		while(testNode) {
			int nodeX = testNode->t->getCoord().first;
			int nodeY = testNode->t->getCoord().second;
			std::cout << '(' << nodeX << ',' << nodeY << ')' << ' ';
			testNode = testNode->prev;
		}
		std::cout << "End chain" << std::endl;

		// checks if reached destination
		std::tie(curX, curY) = t->getCoord();
		if(curX == desX && curY == desY) {
			return backTrack(*nodePtr);
		}
		
		neighbours = t->getNeighbours();
		//std::cout << neighbours.size() << std::endl;
		for(int i = 0; i < 9; ++i) {
			tile = neighbours.at(i);

			// checks if tile is nullptr or in closed set
			if(!tile) continue;			
			if(std::find_if(closed.begin(), closed.end(),
				[&tile](const std::unique_ptr<OpenElement> &e) {
					return e->t == tile;
				}) != closed.end()) continue;

			std::tie(childX, childY) = tile->getCoord();
			if(!tile->traversable() && !(childX == desX && childY == desY)) continue; // destination is not traversable

			std::cout << "DIR: " << inverseMapping(i) << std::endl;
			if(tile) std::cout << *tile << std::endl;
			else std::cout << '0' << std::endl;

			std::tie(X, Y) = tile->getCoord();
			std::cout << '(' << X << ',' << Y << ')' << std::endl;
			std::getline(std::cin, s);			

			gScoreChild = gScore + getCost(i);
			std::tie(childRelX, childRelY) = KEYMAPPING[i];
			fScoreChild = gScoreChild + funcH(childRelX + curX, childRelY + curY, desX, desY);

			auto it = std::find_if(open.begin(), open.end(), 
				[&tile](std::unique_ptr<OpenElement> &tup) {
					return tup->t == tile;
				});
			if(it != open.end() and gScoreChild > (*it)->gScore) {
				// if found, and if gScoreChild >= g-score of accessed element, then continue
				continue;
			}

			int num = i;
			if(num > 3) --num;

			insertSorted(open, fScoreChild, gScoreChild, tile, nodePtr, num);
			std::cout << '(' << X << ',' << Y << ')' << " inserted into open:" << fScoreChild << ' ' << gScoreChild << ' ' << i << std::endl;
			int nodeX = nodePtr->t->getCoord().first;
			int nodeY = nodePtr->t->getCoord().second;
			std::cout << "> child of :" << '(' << nodeX << ',' << nodeY << ')' << std::endl;
		}
	}
	return {};
}

void PathFinder::insertSorted(OpenContainer &open, int fScore, int gScore,
	Tile *t, OpenElement *prev, int dir) {
	auto oe = std::make_unique<OpenElement>(fScore, gScore, t, prev, dir);
	for(auto it = open.begin(); it != open.end(); ++it) {
		if(fScore <= (*it)->fScore) {
			open.insert(it, std::move(oe));
			return;
		}
	}
	open.push_back(std::move(oe));
}

int PathFinder::getCost(int i) {
	if(i == 0 || i == 2 || i == 5 || i == 7) {
		return DIAG_COST;
	}
	return ORTHO_COST;
}

std::queue<int> PathFinder::backTrack(OpenElement &oe) {
	std::cout << "backtracking" << std::endl;
	std::stack<int> reversed;
	std::queue<int> directions;

	reversed.push(oe.dir);
	OpenElement *node = oe.prev;
	while(node) {
		reversed.push(node->dir);
		std::cout << node->dir << std::endl;
		node = node->prev;
	}

	while(!reversed.empty()) {
		directions.push(reversed.top());
		reversed.pop();
	}
	directions.pop(); // remove DEFAULT value
	return directions;
}

File: search/pathFinder.py
from utils.constants import *
from collections import deque
from utils.priority_queue import PriorityQueue

FSCORE = 0
GSCORE = 1
CUR_TILE = 2
PREV_TUP = 3
DIR = 4


class PathFinder:
    __doc__ = '''
    Class that implements the A* search algorithm that integrates with the simulation mechanics
    '''

    def __init__(self, ground, heuristic):
        '''
        Constructor for PathFinder

        :param ground: np.ndarray - 2D array of tiles
        :param heuristic: (Int, Int, Int, Int) => Int - cost function as a heuristic function to be minimized
        '''
        self.ground = ground
        self.heuristic = heuristic

    def find_path(self, x, y, desX, desY, car):
        '''
        Runs the A* search algorithm using the heuristic function.

        :param x: Int - current x-coordinate
        :param y: Int - current y-coordinate
        :param desX: Int - destination x-coordinate
        :param desY: Int - destination y-coordinate
        :param car: Car - reference to Car object
        :return: deque/None - return a deque object containing directions in order if a
                              route exists, otherwise None
        '''
        heuristic = self.heuristic

        # format:             fScore,          gScore, current tile, previous tuple, direction
        #                       0                 1          2              3            4
        first_node = (heuristic(x, y, desX, desY), 0, car.cur_road, None, DEFAULT)
        open = PriorityQueue()
        open.push(first_node)  # begins with the original node
        closed = set()

        while not open.empty():
            node = open.pop()
            fScore, gScore, tile, prev, dir = node
            closed.add(tile)

            curX, curY = tile.get_coord()
            if curX == desX and curY == desY:  # destination reached!
                return self.backtrack(node)

            neighbours = tile.neighbours
            for i in range(9):
                tile_neighbour = neighbours[i]

                # skip None
                if not tile_neighbour: continue
                # skip tiles already considered
                if tile_neighbour in closed: continue

                childX, childY = tile_neighbour.get_coord()
                # skip if tile is not traversable (and not the destination)
                if not tile_neighbour.traversable() and not (childX == desX and childY == desY): continue

                if i > 3: i -= 1  # calibration due to discrepancy

                # builds upon the cost from the original tile
                gScore_child = gScore + self.get_cost(i)
                # gets relative direction from the original tile
                child_relX, child_relY = KEYMAPPING[i]
                fScore_child = gScore_child + heuristic(child_relX + curX, child_relY + curY, desX, desY)

                # finds results if exists
                find_result = open.find(tile_neighbour, key=lambda t: t[CUR_TILE])
                if find_result and gScore_child >= find_result[GSCORE]:
                    # skips if results exists and has a higher g-score
                    continue

                # push into priority queue
                open.push((fScore_child, gScore_child, tile_neighbour, node, i))

        return None  # no route can be found

    def get_cost(self, i):
        '''
        Performs a mapping of direction to costs

        :param i: Int - direction
        :return: Int - cost for the direction
        '''
        if i == NW or i == NE or i == SW or i == SE:
            return DIAG_COST
        return ORTHO_COST

    def backtrack(self, node):
        '''
        Recursively traces back starting from the initial end node that reached the destination
        to obtain a linear forward order of directions to take.

        :param node: tuple(Int, Int, Traversable, tuple(node)/None, Int) - final node containing all previous nodes
        :return: deque - ordered queue of directions to take
        '''
        directions = deque()

        directions.append(node[PREV_TUP])
        while node[PREV_TUP]:
            directions.append(node[DIR])
            node = node[PREV_TUP]

        directions.reverse()
        directions.pop()  # removes the base case recursion

        return directions
